homeassistant/components/sensor/tof.py

```diff
@@ -0,0 +1,115 @@
+"""
+Platform for tof
+Time of Flight - VL53L1X Laser Ranger.
+
+For more details about this platform, please refer to
+https://github.com/josemotta/vl53l1x-python
+
+Fixed setup for current driver version:
+
+- DEFAULT_RANGE is always LONG
+- A GPIO connected to VL53L1X XSHUT input resets the device.
+- XSHUT starts pulsing LOW and after that it is kept HIGH all time.
+
+"""
+from datetime import timedelta
+from functools import partial
+import logging
+import time
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.components import rpi_gpio
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import CONF_NAME
+from homeassistant.helpers.entity import Entity
+
+REQUIREMENTS = ['smbus2==0.2.2', 'VL53L1X2==0.1.5']
+
+DEPENDENCIES = ['rpi_gpio']
+
+_LOGGER = logging.getLogger(__name__)
+
+LENGTH_MILIMETERS = 'mm'
+
+CONF_I2C_ADDRESS = 'i2c_address'
+CONF_I2C_BUS = 'i2c_bus'
+CONF_RANGE = 'range'
+CONF_XSHUT = 'xshut'
+
+DEFAULT_NAME = 'VL53L1X'
+DEFAULT_I2C_ADDRESS = 0x29
+DEFAULT_XSHUT = 16
+DEFAULT_SENSOR_ID = 123
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=5)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_NAME,
+                 default=DEFAULT_NAME): cv.string,
+    vol.Optional(CONF_I2C_ADDRESS,
+                 default=DEFAULT_I2C_ADDRESS): vol.Coerce(int),
+    vol.Optional(CONF_XSHUT,
+                 default=DEFAULT_XSHUT): cv.positive_int,
+})
+
+
+async def async_setup_platform(hass,
+                               config,
+                               async_add_entities,
+                               discovery_info=None):
+    """Setup the VL53L1X ToF Sensor from ST."""
+    from VL53L1X2 import VL53L1X  # pylint: disable=import-error
+
+    name = config.get(CONF_NAME)
+    unit = LENGTH_MILIMETERS
+    xshut = config.get(CONF_XSHUT)
+
+    #  pulse XSHUT port and keep it HIGH
+    rpi_gpio.setup_output(xshut)
```

Comment:
Is this async safe, ie no I/O or blocking?

Reply:
Not sure about that. The intention is keep the port LOW for a few milliseconds to reset the device and then keep it HIGH all the remaining time. I´m testing this on RPI3 and RPI0W. apparently with no blocking issues.

Reply:
If it's not a pure memory operation it needs to be scheduled on the executor thread pool.

Reply:
You mean this? Should I issue a separate schedule for each rpi_gpio operation?

    await hass.async_add_executor_job(
        rpi_gpio.setup_output(xshut)
    )

Reply:
Target function and arguments are passed separately to `hass.async_add_executor_job`. Look at the method signature in core.py.

If we can group function calls that should be run sequentially on the executor thread pool in one function it's good to do that, and just schedule that function. Less overhead that way.

Reply:
I got it, tests running fine. I will commit.

```
    def measure(self):
        """Get the latest measurement from VL53L1X."""
        self.vl53l1x_sensor.start_ranging(self.i2c_address, 2)
        self.vl53l1x_sensor.update(self.i2c_address)
        self.vl53l1x_sensor.stop_ranging(self.i2c_address)

    async def async_update(self):
        """Get the latest measurement and update state."""

        await self.hass.async_add_executor_job(
            self.measure
        ) ...
```

Reply:
These calls are still done from async context.

Reply:
Sorry, please elaborate.

Reply:
I'm assuming `rpi_gpio.setup_output(xshut)` does I/O. So it, and all other similar calls using `rpi_gpio` methods, needs to be scheduled on the executor thread pool with `hass.async_add_executor_job`.

Reply:
Yes, now I got it. I missed the initialization. Is this fine?

```
def init_tof(xshut, level):
    """XSHUT port LOW resets the device"""
    rpi_gpio.setup_output(xshut)
    rpi_gpio.write_output(xshut, level)

async def async_setup_platform(hass,
                               config,
                               async_add_entities,
                               discovery_info=None):
    """Reset and initialize the VL53L1X ToF Sensor from STMicroelectronics."""
    from VL53L1X2 import VL53L1X  # pylint: disable=import-error
...
    xshut = config.get(CONF_XSHUT)

    await hass.async_add_executor_job(
        init_tof, xshut, 0
    )
    await asyncio.sleep(0.01)
    await hass.async_add_executor_job(
        init_tof, xshut, 1
    )
    await asyncio.sleep(0.01)

    sensor = await hass.async_add_executor_job(
        partial, VL53L1X, bus_number
    )
```